internal/logging/sentry_writer/writer.go

```diff
@@ -0,0 +1,262 @@
+	apply(*config)
+}
+
+type optionFunc func(*config)
+
+func (fn optionFunc) apply(c *config) { fn(c) }
+
+type config struct {
+	levels       []zerolog.Level
+	sampleRate   float64
+	release      string
+	environment  string
+	serverName   string
+	ignoreErrors []string
+	debug        bool
+	flushTimeout time.Duration
+}
+
+// WithLevels configures zerolog levels that have to be sent to Sentry.
+// Default levels are: error, fatal, panic.
+func WithLevels(levels ...zerolog.Level) WriterOption {
+	return optionFunc(func(cfg *config) {
+		cfg.levels = levels
+	})
+}
+
+// WithSampleRate configures the sample rate as a percentage of events to be sent in the range of 0.0 to 1.0.
+func WithSampleRate(rate float64) WriterOption {
+	return optionFunc(func(cfg *config) {
+		cfg.sampleRate = rate
+	})
+}
+
+// WithRelease configures the release to be sent with events.
+func WithRelease(release string) WriterOption {
+	return optionFunc(func(cfg *config) {
+		cfg.release = release
+	})
+}
+
+// WithEnvironment configures the environment to be sent with events.
+func WithEnvironment(environment string) WriterOption {
+	return optionFunc(func(cfg *config) {
+		cfg.environment = environment
+	})
+}
+
+// WithServerName configures the server name field for events. Default value is OS hostname.
+func WithServerName(serverName string) WriterOption {
+	return optionFunc(func(cfg *config) {
+		cfg.serverName = serverName
+	})
+}
+
+// WithIgnoreErrors configures the list of regexp strings that will be used to match against event's message
+// and if applicable, caught errors type and value. If the match is found, then a whole event will be dropped.
+func WithIgnoreErrors(reList []string) WriterOption {
+	return optionFunc(func(cfg *config) {
+		cfg.ignoreErrors = reList
+	})
+}
+
+// WithDebug enables sentry client debug logs.
+func WithDebug() WriterOption {
+	return optionFunc(func(cfg *config) {
+		cfg.debug = true
+	})
+}
+
+// New creates writer with provided DSN and options.
+func New(dsn string, opts ...WriterOption) (*Writer, error) {
+	cfg := newDefaultConfig()
+	for _, opt := range opts {
+		opt.apply(&cfg)
+	}
+
+	err := sentry.Init(sentry.ClientOptions{
+		Dsn:          dsn,
+		SampleRate:   cfg.sampleRate,
+		Release:      cfg.release,
+		Environment:  cfg.environment,
+		ServerName:   cfg.serverName,
+		IgnoreErrors: cfg.ignoreErrors,
+		Debug:        cfg.debug,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("cannot initialize sentry: %w", err)
+	}
+
+	levels := make(map[zerolog.Level]struct{}, len(cfg.levels))
+	for _, lvl := range cfg.levels {
+		levels[lvl] = struct{}{}
+	}
+
+	return &Writer{
+		client:       sentry.CurrentHub().Client(),
+		levels:       levels,
+		flushTimeout: cfg.flushTimeout,
+	}, nil
+}
+
+func newDefaultConfig() config {
+	return config{
+		levels: []zerolog.Level{
+			zerolog.ErrorLevel,
+			zerolog.FatalLevel,
+			zerolog.PanicLevel,
+		},
+		sampleRate:   1.0,
+		flushTimeout: 3 * time.Second,
+	}
+}
```

Comment:
Why are you making a copy? I mean, it is fine to simply use the library if you want to use it "as is". I was thinking loud in the other PR that maybe creating these sentry events via zerolog hook could be so simple that we would not need any dependancy at all.

If you really want to use the parsing approach, then just remove this copy and use the library.

Reply:
As I explained offline, I needed https://github.com/archdx/zerolog-sentry/pull/5, but I'll try to implement it with a hook and let see, I agree that if that works, we are bit more efficient and we will need no dependency :+1: